### DISCUSSION TITLE
Is there a method to convert the substructures within Morgan fingerprints into molecules?

### DISCUSSION THREAD

**Discussion by Rainsmumu** [rdkit/rdkit#7031] [Q&A]
I have been using MPNN for molecular property prediction tasks recently. Based on my research idea, I now need to encode the substructures contained in Morgan fingerprints using an MPN encoder. However, the problem is that many of the substructures represented by each position in the Morgan fingerprint are invalid molecules. For example, if there is a molecule with a benzene ring, it is highly probable that there will be broken benzene rings in its substructures. Undoubtedly, these substructures cannot be input into the MPNN. So, I'm wondering if there is a method to make these substructures valid and treat them as complete molecules. Currently, I have two ideas:
1. If the substructure contains 'c', replace it with the entire benzene ring (a rough idea).
2. Remove all invalid substructures and only keep those that can be converted into molecules.

Any suggestions would be greatly appreciated!

**Comment by aish-tan** [comment]
Hi,

While I have negligible experience with MPNN, based on this tutorial (https://keras.io/examples/graph/mpnn-molecular-graphs/), I don't see Morgan Fingerprints (MF) being used in this method at all. Hence, I can't help you there. 
However, MF when used as bits, can be used as valid model inputs. Way to do that in RDKit is:
```
fp_gen = rdFingerprintGenerator.GetMorganGenerator(radius=radius, fpSize=n_bits, includeChirality=use_chirality)
```
More can be read in Greg's blog: https://greglandrum.github.io/rdkit-blog/posts/2023-01-18-fingerprint-generator-tutorial.html

Later on, if you want to visualize the substructures encoded by the MF bits in your dataset, you can use X-FP:
https://github.com/czodrowskilab/x-fp

Hope this helps.

**Comment by Rainsmumu** [comment]
@aish-tan Hi! Thank you so much for your prompt response. I will carefully review the relevant documents and links you shared.

**Comment by dehaenw** [comment]
hi @Rainsmumu 
to do this you can use FindAtomEnvironmentOfRadiusN and MolFragmentToSmiles. The trick to make them parsable is to turn on the flag kekuleSmiles. in case of a partial aromatic ring it will give a smiles like "C:C:C:C:C" which is parsable.

here's an example codeblock where all smiles corresponding to morgan bits are printed and their molecules displayed:
```python
from rdkit import Chem
from rdkit.Chem import rdFingerprintGenerator

mol=Chem.MolFromSmiles(Chem.MolToSmiles(Chem.MolFromSmiles("c1[nH]osc1CO")))

def get_morgan_bit_smi(mol,rootatom,radius):
    if radius>0:
        bit_atoms=Chem.FindAtomEnvironmentOfRadiusN(mol,radius,rootatom)
        amap={}
        submol=Chem.PathToSubmol(mol,bit_atoms,atomMap=amap)
        smi = Chem.MolFragmentToSmiles(mol,atomsToUse=amap,kekuleSmiles=True,allBondsExplicit=True,allHsExplicit=True)
    else:
        smi = Chem.MolFragmentToSmiles(mol,atomsToUse=(rootatom,),kekuleSmiles=True,allBondsExplicit=True,allHsExplicit=True)
    return smi

mfpgen = rdFingerprintGenerator.GetMorganGenerator(radius=3)
ao = rdFingerprintGenerator.AdditionalOutput()
ao.AllocateBitInfoMap()
fp = mfpgen.GetFingerprint(mol,additionalOutput=ao)
bi=ao.GetBitInfoMap()
for bit in bi:
    for b in bi[bit]:
        smi=get_morgan_bit_smi(mol,b[0],b[1])
        print(f'bit {bit} corresponds to {smi}')
        display(Chem.MolFromSmiles(smi))
```

**Reply by Rainsmumu** [reply]
Thank you for the help. I appreciate it.@dehaenw 
the codeblock you provided is very helpful, now I can input these substructures into MPNN.
Just one more question:
why kekuleSmiles can be parsed while Smiles not, isn't the difference between them just in the different forms of representing bonds within aromatic rings?
